0-module-and-module-source.emu

```diff
@@ -0,0 +1,650 @@
+              ): ~unused~
+            </h1>
+            <dl class="header">
+              <dt>description</dt>
+              <dd>FinishModuleDynamicImport completes the process of a dynamic import originally started by an <emu-xref href="#sec-import-calls">`import()`</emu-xref> call for a Module Instance, resolving or rejecting the promise returned by that call as appropriate according to _innerPromise_'s resolution. It is performed by host environments as part of HostImportModuleRecordDynamically.</dd>
+            </dl>
+            <emu-alg>
+              1. Let _fulfilledClosure_ be a new Abstract Closure with parameters (_result_) that captures _moduleRecord_ and _promiseCapability_ and performs the following steps when called:
+                1. Assert: _result_ is *undefined*.
+                1. Assert: Evaluate has already been invoked on _moduleRecord_ and successfully completed.
+                1. Let _namespace_ be Completion(GetModuleNamespace(_moduleRecord_)).
+                1. If _namespace_ is an abrupt completion, then
+                  1. Perform ! Call(_promiseCapability_.[[Reject]], *undefined*, &laquo; _namespace_.[[Value]] &raquo;).
+                1. Else,
+                  1. Perform ! Call(_promiseCapability_.[[Resolve]], *undefined*, &laquo; _namespace_.[[Value]] &raquo;).
+                1. Return ~unused~.
+              1. Let _onFulfilled_ be CreateBuiltinFunction(_fulfilledClosure_, 0, *""*, &laquo; &raquo;).
+              1. Let _rejectedClosure_ be a new Abstract Closure with parameters (_error_) that captures _promiseCapability_ and performs the following steps when called:
+                1. Perform ! Call(_promiseCapability_.[[Reject]], *undefined*, &laquo; _error_ &raquo;).
+                1. Return ~unused~.
+              1. Let _onRejected_ be CreateBuiltinFunction(_rejectedClosure_, 0, *""*, &laquo; &raquo;).
+              1. Perform PerformPromiseThen(_innerPromise_, _onFulfilled_, _onRejected_).
+              1. Return ~unused~.
+            </emu-alg>
+
+            <emu-note type="editor">
+                This Abstract Operation is analogous to FinishDynamicImport but for Module Records with an associated Module Instance.
+            </emu-note>
+          </emu-clause>
+    </emu-clause>
+
+    <emu-clause id="sec-module-constructor">
+        <h1>The Module Constructor</h1>
+        <p>The Module constructor:</p>
+        <ul>
+            <li>is the intrinsic object <dfn>%Module%</dfn>.</li>
+            <li>is the initial value of the *"Module"* property of the global object.</li>
+            <li>is not intended to be called as a function and will throw an exception when called in that manner.</li>
+            <li>creates and initializes a new Module object when called as a constructor.</li>
+        </ul>
+
+        <emu-clause id="sec-module">
+            <h1>Module ( _moduleSource_, _referral_, _importHook_ [, _importMeta_ ] )</h1>
```

Comment:
We have always discussed about a _string_ referrer, but this spec text allows `referral` to be any ECMAScript value. I very much prefer this behavior, rather than restricting to strings (I was going to open a new issue proposing to relax the restriction, before noticing that it's already not restricted).

I realized that passing the referrer to `importHook` (and thus passing `referral` to `new Module`) is not necessary, but it's just an ergonomic improvement. If the spec didn't call `importHook` passing the referrer, I could just do this:
```js
function buildImportHook(referrer) {
  return async function importHook(specifier) {
    const url = new URL(specifier, referrer);
    const sourceText = await fetch(url).then(r => r.text());
    const source = new ModuleSource(sourceText);
    return new Module(source, buildImportHook(url));
  };
};

const entryModule = new Module(entrySource, buildImportHook(entryURL));
await import(entryModule);
```
(or I could use `.bind` rather than manual currying)

Passing the `referrer` through the `Module` constructor to `importHook` is just an ergonomic/performance improvement to avoid creating closures that capture context: it's similar to the `thisArg` parameter of all the array methods that take a function.

Since we are "just" improving ergonomics, we can look at other use cases for providing some context to the `importHook` function. One example ([taken from the HTML spec](https://html.spec.whatwg.org/#script-fetch-options)) is to provide some "fetch options" that describe how the dependencies should be fetched:
```js
async function importHook(specifier, { baseURL, referrerPolicy }) {
  const url = new URL(specifier, baseURL);
  const sourceText = await fetch(url, { referrerPolicy }).then(r => r.text());
  const source = new ModuleSource(sourceText);
  return new Module(source, {
    baseURL: url,
    referrerPolicy,
  });
}

const entryModule = importHook("./main.js", { baseURL: document.URL, referrerPolicy: "same-origin" });
await import(entryModule);

// That is roughly equivalent to
//   <script type="module" src="./main.js" referrerpolicy="same-origin" />
```

Allowing any object aligns this "context parameter" with the current [[HostDefined]] slot on Module Records: it's a general-purpouse place used to share loader-level information between modules.

So :+1: to not checking that `Type(_referral_)` is string, and maybe we could rename it to `_context_` (or `_thisArg_` and pass it as `this` to `importHook`, like array methods do).

Reply:
@nicolo-ribaudo  in principle I agree with this. Question: in the example above I suspect you are missing `importHook`  argument when calling `new Module()` right?

Reply:
Yes, fixed!

Reply:
Resolution: We will look into this in another PR.